src/main/kotlin/org/jetbrains/research/tasktracker/tracking/activity/ActivityTracker.kt

```diff
@@ -0,0 +1,123 @@
+
+    // TODO: add config to select activities to track
+    fun startTracking() {
+        trackingDisposable = Disposer.newDisposable()
+        listenActions()
+        listenKeyboard()
+        listenExecution()
+    }
+
+    fun stopTracking() {
+        messageBusConnections.forEach { it.disconnect() }
+        trackingDisposable?.let { Disposer.dispose(it) }
+    }
+
+    private fun listenActions(listenShortcut: Boolean = true) {
+        val actionListener = object : AnActionListener {
+            override fun beforeActionPerformed(anAction: AnAction, event: AnActionEvent) {
+                val actionId = ActionManager.getInstance().getId(anAction)
+                    ?: anAction.javaClass.name
+                val id = currentId.getAndIncrement()
+                if (listenShortcut) {
+                    actionId.extractShortcut(event, id)
+                }
+                activityLogger.log(Type.Action, actionId, id)
+            }
+        }
+        actionListener.connect(AnActionListener.TOPIC)
+    }
+
+    private fun String.extractShortcut(event: AnActionEvent, id: Int) {
+        if (!movingActions.contains(this) && !typingActions.contains(this)) {
+            val input = event.inputEvent
+            if (input is KeyEvent) {
+                activityLogger.log(Type.Shortcut, eventToShortcutInfo(input), id)
+            }
+        }
+    }
+
+    private fun eventToShortcutInfo(event: KeyEvent): String {
+        val keys = Key.values().mapNotNull { it.check(event) }.map { it.name }.plus(event.keyCode.toString())
+        return keys.joinToString(" ")
+    }
+
+    private fun listenExecution() {
+        val executionListener = object : ExecutionListener {
+            override fun processStarting(executorId: String, env: ExecutionEnvironment, handler: ProcessHandler) {
+                val commandLine = (handler as? BaseProcessHandler<*>)?.commandLine ?: ""
+                activityLogger.log(Type.Execution, "$executorId:'${env.runProfile}':$commandLine")
+            }
+        }
+        executionListener.connect(ExecutionManager.EXECUTION_TOPIC)
+    }
+
+    private fun listenKeyboard() {
+        IdeEventQueue.getInstance().addPostprocessor({ awtEvent: AWTEvent ->
+            if (awtEvent is KeyEvent && awtEvent.id == KeyEvent.KEY_PRESSED) {
+                activityLogger.log(Type.KeyPressed, awtEvent.info())
+            }
+            if (awtEvent is KeyEvent && awtEvent.id == KeyEvent.KEY_RELEASED) {
+                activityLogger.log(Type.KeyReleased, awtEvent.info())
+            }
+            if (awtEvent is KeyEvent) {
+                when (awtEvent.id) {
+                    KeyEvent.KEY_PRESSED -> activityLogger.log(Type.KeyPressed, awtEvent.info())
+                    KeyEvent.KEY_RELEASED -> activityLogger.log(Type.KeyReleased, awtEvent.info())
+                }
+            }
+            false
+        }, trackingDisposable)
+    }
+
+    private fun <L : Any> L.connect(topic: Topic<L>) {
```

Comment:
Should we change Any to Listener?

Reply:
No, since there are no super class for all listeners